module/jobqueue/ready_done_aware_consumer.go

```diff
@@ -0,0 +1,126 @@
+package jobqueue
+
+import (
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/storage"
+)
+
+type ReadyDoneAwareConsumer struct {
+	component.Component
+
+	cm         *component.ComponentManager
+	consumer   module.JobConsumer
+	jobs       module.Jobs
+	workSignal <-chan struct{}
+	notifier   NotifyDone
+	log        zerolog.Logger
+}
+
+// NewReadyDoneAwareConsumer creates a new ReadyDoneAwareConsumer consumer
+func NewReadyDoneAwareConsumer(
+	log zerolog.Logger,
+	progress storage.ConsumerProgress,
+	jobs module.Jobs,
+	processor JobProcessor, // method used to process jobs
+	workSignal <-chan struct{},
+	defaultIndex uint64,
+	maxProcessing uint64,
+	maxSearchAhead uint64,
+	notifier NotifyDone,
+) (*ReadyDoneAwareConsumer, error) {
+
+	c := &ReadyDoneAwareConsumer{
+		workSignal: workSignal,
+		notifier:   notifier,
+		jobs:       jobs,
+		log:        log,
+	}
+
+	// create a worker pool with maxProcessing workers to process jobs
+	worker := NewWorkerPool(
+		processor,
+		func(id module.JobID) { c.NotifyJobIsDone(id) },
+		maxProcessing,
+	)
+	c.consumer = NewConsumer(c.log, c.jobs, progress, worker, maxProcessing, maxSearchAhead)
+
+	builder := component.NewComponentManagerBuilder().
+		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			worker.Start(ctx)
+			<-worker.Ready()
+
+			c.log.Info().Msg("job consumer starting")
+			err := c.consumer.Start(defaultIndex)
+			if err != nil {
+				ctx.Throw(fmt.Errorf("could not start consumer: %w", err))
+			}
+
+			ready()
+
+			<-ctx.Done()
+			c.log.Info().Msg("job consumer shutdown started")
+
+			// blocks until all running jobs have stopped
+			c.consumer.Stop()
+
+			<-worker.Done()
+			c.log.Info().Msg("job consumer shutdown complete")
+		}).
+		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			ready()
+			c.processingLoop(ctx)
+		})
+
+	cm := builder.Build()
+	c.cm = cm
+	c.Component = cm
+
+	return c, nil
+}
+
+// NotifyJobIsDone is invoked by the worker to let the consumer know that it is done
+// processing a (block) job.
+func (c *ReadyDoneAwareConsumer) NotifyJobIsDone(jobID module.JobID) uint64 {
```

Comment:
this would mean the `ReadyDoneAwareConsumer` no longer implements to `JobConsumer` interface. I think it would be better to continue implementing it for now, at least until we convert the other jobs to use the ready done aware interface.